3-make_gene_matrix.R

```diff
@@ -0,0 +1,147 @@
+salmon.folders <- dir(opt$dir) 
+
+# Read the data into a list
+salmon <- lapply(salmon.folders, function(x) {
+                read.table(file.path(opt$dir, x, "quant.sf"), header = TRUE)
+                })
+
+# Get Salmon ensembl gene annotation IDs
+salmon.annot <- strsplit(as.character(salmon[[1]]$Name), "\\|")
+salmon.annot <- vapply(salmon.annot, function(x) {
+                        x <- grep("ENST", x, value = TRUE)
+                             gsub("\\.[0-9]+$", "", x)
+                          }, FUN.VALUE = "character")
+
+# Turn transcript ids to gene ids
+salmon.annot <- mapIds(org.Hs.eg.db, keys = salmon.annot, column = "ENSEMBL", 
+                       keytype = "ENSEMBLTRANS")
+
+# Make a matrix of the data
+salmon.data <- do.call("cbind", lapply(salmon, function(x) {
+                          tapply(x$NumReads, salmon.annot, sum)
+                          }))
+
+# Carry over ID names
+dimnames(salmon.data)[[2]] <- salmon.folders
+
+# Make into a make annotation into a dataframe (removes data without a gene)
+salmon.data <- data.frame('ensembl' = rownames(salmon.data),
+                          'gene' = mapIds(org.Hs.eg.db, keys = rownames(salmon.data), 
+                                          column = "SYMBOL", keytype = "ENSEMBL"),
+                           salmon.data,
+                           stringsAsFactors = FALSE)
+
+#---------------------Salmon proportion of mapped reads------------------------#
+# Get the proportion of mapped reads
+salmon.prop.assigned <- vapply(salmon.folders, function(x) {
+  rjson::fromJSON(file = file.path(opt$dir, x, "aux_info",
+                                   "meta_info.json"))$percent_mapped/100
+  }, FUN.VALUE = 1)
+
+# Make a histogram of this information
+png(file.path(opt$output, "salmon_prop_reads_mapped_hist.ensembl.png"))
+hist(salmon.prop.assigned, xlab = "", main = "Salmon Proportion of Mapped Reads",
+breaks = 20)
+dev.off()
+
+# Filter out samples with too low of mapped reads 
+salmon.data <- salmon.data[, which(salmon.prop.assigned > opt$mapped)]
+
+#--------------------------- Create ID conversion key--------------------------#
+if (!file.exists(file.path(opt$output, "sample_id_key.RDS"))) {
+    # Get geo metadata
+    geo.meta <- GEOquery::getGEO(opt$geo, destdir = file.path(opt$output))
+    
+    # Get the GSM and SRX ids from the GEO metadata
+    id.key <- data.frame(gsm.ids = geo.meta[[1]]@phenoData@data$geo_accession,
+    plate.ids = geo.meta[[1]]@phenoData@data$description.1,
+    srx.ids = stringr::word(geo.meta[[1]]@phenoData@data$relation.1,
+    start = 2, sep = "term="))
+    
+    # Import SRA file names
+    sra.files <- read.csv(file.path(opt$output, "SRA.files.csv"))[, -1]
+    
+    # Merge both keys into a single id.key dataframe
+    id.key <- merge(id.key, sra.files, by.x = "srx.ids", by.y = "experiment")
+    
+    # Save this dataframe for later
+    saveRDS(id.key, file = file.path(opt$output, "sample_id_key.RDS"))
```

Comment:
If this is tabular data, why save it as a `.RDS`?

Reply:
I guess because the next scripts are just going to reload the data into the R environment so this seemed like the easiest way to keep the object exactly as I want it to be in R. I could write this to a tsv and then use read_tsv. Or I could save it as both an RDS and tsv. IDK what's ideal. I guess tsv in case people want to use the file for something else in the intermediate?